In [1]:
################################################################################
# Definindo um caminho personalizado para bibliotecas R
################################################################################
# Exemplo de caminho: pode ser uma pasta dentro do seu diretório home
# Lembre-se que o R pode adicionar subdiretórios específicos da versão/arquitetura
meu_caminho_biblioteca <- "~/R/library"

# Define a variável de ambiente R_LIBS_USER para a sessão atual
Sys.setenv(R_LIBS_USER = meu_caminho_biblioteca)

# Verifica o valor da variável de ambiente
Sys.getenv("R_LIBS_USER")

# Mostra todos os caminhos de biblioteca que o R está usando
# O seu novo caminho deve aparecer aqui, geralmente como o primeiro da lista.
.libPaths()

[1] "~/R/library"

[1] "/usr/local/lib/R/site-library" "/usr/local/lib/R/library"     
[3] "/home/vscode/R/library"

In [2]:
################################################################################
#               INSTALAÇÃO E CARREGAMENTO DE PACOTES NECESSÁRIOS               #
################################################################################

# Pacotes utilizados
pacotes <- c("tidyverse", #carregar outros pacotes do R
             "knitr", "kableExtra", #formatação de tabelas
             "arrow","duckdb") # parquet e banco de dados

if(sum(as.numeric(!pacotes %in% installed.packages())) != 0){
  instalador <- pacotes[!pacotes %in% installed.packages()]
  for(i in 1:length(instalador)) {
    install.packages(instalador, dependencies = T, lib = meu_caminho_biblioteca)
    break()}
  sapply(pacotes, require, character = T, lib = meu_caminho_biblioteca) 
} else {
  sapply(pacotes, require, character = T, lib = meu_caminho_biblioteca) 
}

Loading required package: tidyverse

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.6.0
✔ ggplot2   4.0.1     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.2.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Loading required package: knitr

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘knitr’”
Loading required package: kableExtra


Attaching package: ‘kableExtra’


The following object is masked from ‘package:dplyr’:

    group_rows


Loading required package: arrow


Attaching package: ‘arrow’


The following object is masked from ‘package:lubridate’:

 

tidyverse      knitr kableExtra      arrow     duckdb 
      TRUE      FALSE       TRUE       TRUE       TRUE

# Carga de dados sobre condições de pavimento das rodovias brasileiras

https://servicos.dnit.gov.br/vgeo/

Shapefile

https://www.geoaplicada.com/dados/rodovias-do-brasil/#:~:text=%C3%89%20poss%C3%ADvel%20baixar%20as%20camadas,ano%202013%20at%C3%A9%20o%20presente.&text=Tamb%C3%A9m%20%C3%A9%20poss%C3%ADvel%20baixar%20as,Geogr%C3%A1ficas%20do%20DNIT%20(VGEO).&text=As%20linhas%20de%20rodovias%20estaduais,Contate%2Dnos%20por%20aqui.

https://servicos.dnit.gov.br/dnitcloud/index.php/s/oTpPRmYs5AAdiNr?path=%2FSNV%20Bases%20Geom%C3%A9tricas%20(2013-Atual)%20(SHP)


Como avaliar as condições de pavimentação das rodovias brasileiras?
- Comparativo rede pavimentada e não pavimentada? (SNV Planilhas)


# Índices de Condição das Rodovias

https://servicos.dnit.gov.br/dadosabertos/dataset/condicoes-do-pavimento

ICC Numérico NN,NNN Índice de Condição da Conservação.
ICP Numérico NN,NNN Índice de Condição do Pavimento
ICM Numérico NN,NNN Índice de Condição de Manutençao, sendo só seguintes valores de referência:
ICM < 30 = BOM
30 < ICM < 50 = REGULAR
50 < ICM < 70 = RUIM
ICM > 70 = PÉSSIMO


## DICIONÁRIO DE DADOS de condições do pavimento -

ÍNDICE DE CONSERVAÇÃO E MANUTENÇÃO - ICM

| Campo | Tipo | Formato | Descrição |
|---|---|---|---|
| Contrato | Alfanumérico | NN NNNNN/NNNN | Refere-se ao número do contrato de serviço |
| UF | Texto | XX | Refere-se ao Estado da Federação que o equipamento está instalado. Indicado pela sigla da Unidade em formato XX. Ex: AL |
| Rodovia | Alfanumérico | XX-NNN | Refere-se a rodovia em análise em formato XX-NNN. Ex: BR-116 |
| Km inicial | Numérico | NNN,NNN | Refere-se ao km da rodovia onde inicia-se a análise, indicado em formato NNN,NNN de número com até 3 algarismos e 3 decimais, separados por vírgula. Ex: 485,140 |
| Km final | Numérico | NNN,NNN | Refere-se ao km da rodovia onde finaliza-se a análise, indicado em formato NNN,NNN de número com até 3 algarismos e 3 decimais, separados por vírgula. Ex: 485,140 |
| Extensão | Numérico | NNN,NNN | Refere-se à diferença entre o km final e o km inicial |
| Data | Data | DD/MM/AAAA | Trata-se da data em que foi realizada a análise mais recente, no formato Dia, Mês, Ano. Ex.: 01/11/2020 |
| Latitude | Numérico | - | Indica a latitude da localização do equipamento, com coordenada geográfica. Indicado em formato XXºXX'XX,XX''X, onde todos são números exceto o último, que indica o hemisfério N (Norte) ou S (Sul). Ex: 13°31'10,91"S |
| Longitude | Numérico | - | Indica a longitude da localização do equipamento, com coordenada geográfica. Indicado em formato XXºXX'XX,XX''O, onde todos são números exceto o último, que indica estar na região Oeste do Meridiano de Greenwich. Ex: 49°11'09,99"O |
| ICC | Numérico | NN,NNN | Índice de Condição da Conservação |
| ICP | Numérico | NN,NNN | Índice de Condição do Pavimento |
| ICM | Numérico | NN,NNN | Índice de Condição de Manutenção. Valores de referência: ICM < 30 = BOM; 30 < ICM < 50 = REGULAR; 50 < ICM < 70 = RUIM; ICM > 70 = PÉSSIMO |


In [ ]:
# Carga de 202010A - Condições do Pavimento em Dezembro/2020
# Monte código R para carga do CSV conforme a seguir. Considere a formatação da data e dos números com vírgula que são floats e estão no formato brasileiro. Padronize os valore nulos para NA.
# Contrato;UF;Rodovia;km inicial;km final;Extensão (km);Data;Latitude;Longitude;ICC;ICP;ICM
# 01 00583/2018;AM;BR-230;213,200;212,000;1,200;30/11/2020;-7,1991363;-59,8909876;80,000;70,000;73,000
# 01 00583/2018;AM;BR-230;212,000;211,000;1,000;30/11/2020;-7,1949552;-59,8855125;80,000;70,000;73,000
# 01 00583/2018;AM;BR-230;211,000;210,000;1,000;30/11/2020;-7,1949552;-59,8855125;80,000;45,000;55,500
# 01 00583/2018;AM;BR-230;210,000;209,000;1,000;30/11/2020;-7,1949552;-59,8855125;80,000;70,000;73,000
# 01 00583/2018;AM;BR-230;209,000;208,000;1,000;30/11/2020;-7,1949552;-59,8855125;80,000;45,000;55,500
# 01 00583/2018;AM;BR-230;213,200;214,000;0,800;30/11/2020;-7,1788378;-59,846564;85,000;70,000;74,500
# 01 00583/2018;AM;BR-230;214,000;215,000;1,000;30/11/2020;-7,1923246;-59,8833152;85,000;70,000;74,500

################################################################################
#          CARGA DE DADOS - CONDIÇÕES DE PAVIMENTO DAS RODOVIAS              #
#                    SNV 202010A - Dezembro/2020                             #
################################################################################

# Carregar bibliotecas necessárias
library(tidyverse)
library(readr)

# Ler o CSV com separador ponto-e-vírgula
# Formato brasileiro: números com vírgula (,) como separador decimal
# Datas no formato DD/MM/YYYY
df_pavimentos <- read.csv2(
  "Trabalho/bases-de-dados/condicoes-de-pavimento/icm-dez20.csv",
  sep = ";",
  encoding = "UTF-8",
  stringsAsFactors = FALSE,
  na.strings = c("", "-", "NA")  # Tratar vazio, hífen e NA como faltante
)

# Exibir dimensões e primeiras linhas
cat("\n=== RESUMO DO CARREGAMENTO ===\n")
cat("Dimensões do conjunto de dados:", 
    nrow(df_pavimentos), "linhas x", 
    ncol(df_pavimentos), "colunas\n\n")

print(paste("A tabela possui", ncol(df_pavimentos), "colunas."))

colunas_com_na <- sum(colSums(is.na(df_pavimentos)) > 0)
print(paste("A tabela possui", colunas_com_na, "colunas com valores ausentes."))

cat("\n=== PRIMEIRAS 6 LINHAS ===\n")
print(head(df_pavimentos, 6))

################################################################################
#                      ESTRUTURA E TIPOS DE DADOS                             #
################################################################################

cat("\n=== ESTRUTURA DOS DADOS (ANTES DA CONVERSÃO) ===\n")
str(df_pavimentos)

################################################################################
#         CONVERSÃO E LIMPEZA DE COLUNAS NUMÉRICAS E DE DATA                  #
################################################################################

# Converter colunas numéricas (formato brasileiro com vírgula)
colunas_numericas <- c(
  "km.inicial", "km.final", "Extensão..km.",
  "ICC", "ICP", "ICM"
)

# Aplicar conversão com tratamento de pontos decimais
for (col in colunas_numericas) {
  if (col %in% names(df_pavimentos)) {
    # Substituir vírgula por ponto e converter para numérico
    df_pavimentos[[col]] <- as.numeric(
      gsub(",", ".", gsub("\\.", "", df_pavimentos[[col]]))
    )
  }
}

# Converter coluna de Latitude e Longitude (também têm vírgula)
df_pavimentos$Latitude <- as.numeric(
  gsub(",", ".", gsub("\\.", "", df_pavimentos$Latitude))
)

df_pavimentos$Longitude <- as.numeric(
  gsub(",", ".", gsub("\\.", "", df_pavimentos$Longitude))
)

# Converter coluna Data de formato DD/MM/YYYY para Date
df_pavimentos$Data <- as.Date(df_pavimentos$Data, format = "%d/%m/%Y")

cat("\n=== ESTRUTURA DOS DADOS (APÓS CONVERSÃO) ===\n")
str(df_pavimentos)



=== RESUMO DO CARREGAMENTO ===
Dimensões do conjunto de dados: 34571 linhas x 12 colunas

[1] "A tabela possui 12 colunas."
[1] "A tabela possui 1 colunas com valores ausentes."

=== PRIMEIRAS 6 LINHAS ===
       Contrato UF Rodovia km.inicial km.final Extensao.km       Data  Latitude
1 01 00583/2018 AM  BR-230      213.2      212         1.2 30/11/2020 -7.199136
2 01 00583/2018 AM  BR-230      212.0      211         1.0 30/11/2020 -7.194955
3 01 00583/2018 AM  BR-230      211.0      210         1.0 30/11/2020 -7.194955
4 01 00583/2018 AM  BR-230      210.0      209         1.0 30/11/2020 -7.194955
5 01 00583/2018 AM  BR-230      209.0      208         1.0 30/11/2020 -7.194955
6 01 00583/2018 AM  BR-230      213.2      214         0.8 30/11/2020 -7.178838
  Longitude ICC ICP  ICM
1 -59.89099  80  70 73.0
2 -59.88551  80  70 73.0
3 -59.88551  80  45 55.5
4 -59.88551  80  70 73.0
5 -59.88551  80  45 55.5
6 -59.84656  85  70 74.5

=== ESTRUTURA DOS DADOS (ANTES DA CONVERSÃO) ===
'data.fr


=== ESTRUTURA DOS DADOS (APÓS CONVERSÃO) ===
'data.frame':	34571 obs. of  12 variables:
 $ Contrato   : chr  "01 00583/2018" "01 00583/2018" "01 00583/2018" "01 00583/2018" ...
 $ UF         : chr  "AM" "AM" "AM" "AM" ...
 $ Rodovia    : chr  "BR-230" "BR-230" "BR-230" "BR-230" ...
 $ km.inicial : num  2132 212 211 210 209 ...
 $ km.final   : num  212 211 210 209 208 214 215 198 199 200 ...
 $ Extensao.km: num  1.2 1 1 1 1 0.8 1 0.3 1 1 ...
 $ Data       : Date, format: "2020-11-30" "2020-11-30" ...
 $ Latitude   : num  -71991363 -71949552 -71949552 -71949552 -71949552 ...
 $ Longitude  : num  -5.99e+08 -5.99e+08 -5.99e+08 -5.99e+08 -5.99e+08 ...
 $ ICC        : num  80 80 80 80 80 85 85 375 30 25 ...
 $ ICP        : num  70 70 45 70 45 70 70 325 25 25 ...
 $ ICM        : num  73 73 555 73 555 745 745 34 265 25 ...


In [14]:

################################################################################
#                    ESTATÍSTICAS DESCRITIVAS                                 #
################################################################################

cat("\n=== ESTATÍSTICAS DESCRITIVAS ===\n")
summary(df_pavimentos)

################################################################################
#                    ANÁLISES EXPLORATÓRIAS INICIAIS                          #
################################################################################

# Distribuição por UF
cat("\n=== DISTRIBUIÇÃO POR UF ===\n")
df_pavimentos %>%
  group_by(UF) %>%
  summarise(
    Total = n(),
    Extensao_Total = sum(Extensao.km, na.rm = TRUE),
    ICC_Medio = mean(ICC, na.rm = TRUE),
    ICP_Medio = mean(ICP, na.rm = TRUE),
    ICM_Medio = mean(ICM, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  arrange(desc(Total)) %>%
  print()

# Distribuição por Rodovia
cat("\n=== DISTRIBUIÇÃO POR RODOVIA ===\n")
df_pavimentos %>%
  group_by(Rodovia) %>%
  summarise(
    Total = n(),
    Extensao_Total = sum(Extensao.km, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  arrange(desc(Total)) %>%
  print()

# Categorização do ICM (Índice de Condição de Manutenção)
cat("\n=== ESTATÍSTICAS DO ICM (Índice de Condição de Manutenção) ===\n")
summary(df_pavimentos$ICM)

# Classificação do ICM
df_pavimentos <- df_pavimentos %>%
  mutate(
    Classificacao_ICM = case_when(
      ICM < 30 ~ "BOM",
      ICM >= 30 & ICM < 50 ~ "REGULAR",
      ICM >= 50 & ICM < 70 ~ "RUIM",
      ICM >= 70 ~ "PÉSSIMO",
      is.na(ICM) ~ NA_character_
    )
  )

cat("\n=== DISTRIBUIÇÃO POR CLASSIFICAÇÃO ICM ===\n")
df_pavimentos %>%
  group_by(Classificacao_ICM) %>%
  summarise(
    Total = n(),
    Extensao_Total = sum(Extensao.km, na.rm = TRUE),
    ICM_Medio = mean(ICM, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  arrange(desc(Total)) %>%
  print()


=== ESTATÍSTICAS DESCRITIVAS ===


   Contrato              UF              Rodovia            km.inicial     
 Length:34571       Length:34571       Length:34571       Min.   :    0.0  
 Class :character   Class :character   Class :character   1st Qu.:   96.0  
 Mode  :character   Mode  :character   Mode  :character   Median :  229.0  
                                                          Mean   :  350.8  
                                                          3rd Qu.:  449.0  
                                                          Max.   :78788.0  
    km.final        Extensao.km           Data               Latitude         
 Min.   :    0.0   Min.   :  0.070   Min.   :2020-11-04   Min.   :-999833166  
 1st Qu.:   96.0   1st Qu.:  1.000   1st Qu.:2020-11-17   1st Qu.:-276137791  
 Median :  227.0   Median :  1.000   Median :2020-11-24   Median :-131302671  
 Mean   :  320.9   Mean   :  1.022   Mean   :2020-11-21   Mean   :-221563090  
 3rd Qu.:  444.0   3rd Qu.:  1.000   3rd Qu.:2020-11-26   3rd Qu.: -58130


=== DISTRIBUIÇÃO POR UF ===
# A tibble: 21 × 6
   UF    Total Extensao_Total ICC_Medio ICP_Medio ICM_Medio
   <chr> <int>          <dbl>     <dbl>     <dbl>     <dbl>
 1 MT     3832          3756.      160.      34.9      929.
 2 MA     3472          3470.      126.     114.      1477.
 3 MS     3057          3036.      174.      99.2     1606.
 4 PI     2805          2783.      215.      66.7     1318.
 5 PE     2358          3293       198.      35.8     1043.
 6 PA     2336          2329.      257.      65.6     1749.
 7 CE     2267          2260       176.     117.      1691.
 8 TO     1920          1921.      181.      28.1      994.
 9 RO     1816          1816.      238.     139.      2025.
10 PB     1572          1557.      187.     141.      2044.
# ℹ 11 more rows

=== DISTRIBUIÇÃO POR RODOVIA ===
# A tibble: 79 × 3
   Rodovia Total Extensao_Total
   <chr>   <int>          <dbl>
 1 BR-364   2943          2901.
 2 BR-316   1950          1943.
 3 BR-174   1915          1896.
 4

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
      6      71     465    1453    2225    9325 


=== DISTRIBUIÇÃO POR CLASSIFICAÇÃO ICM ===


# A tibble: 4 × 4
  Classificacao_ICM Total Extensao_Total ICM_Medio
  <chr>             <int>          <dbl>     <dbl>
1 PÉSSIMO           25939         26753.    1928. 
2 BOM                5671          5623.      16.1
3 REGULAR            2187          2174.      38.8
4 RUIM                774           770       57.5


In [16]:
# Gravar df_pavimento num arquivo parquet "Trabalho/bases-de-dados/condicoes-de-pavimento/
# condicoes-de-pavimento-pq" particionado por ano

################################################################################
#        SALVAR DATASET EM FORMATO PARQUET PARTICIONADO POR ANO               #
################################################################################

# Carregar biblioteca arrow (se não estiver carregada)
library(arrow)

# Preparar dataframe: extrair ano da coluna Data
df_pavimentos <- df_pavimentos %>%
  mutate(
    Ano = format(Data, "%Y")
  )

# Exibir distribuição por ano antes de salvar
cat("\n=== DISTRIBUIÇÃO DO DATASET POR ANO ===\n")
print(table(df_pavimentos$Ano))

# Salvar em Parquet particionado por Ano
write_dataset(
  df_pavimentos,
  path = "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq",
  format = "parquet",
  partitioning = "Ano"
)

cat("\n✓ Dataset salvo com sucesso em formato Parquet\n")
cat("   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/\n")
cat("   Estrutura: Ano=2020/, Ano=2021/, Ano=2022/, etc.\n")



=== DISTRIBUIÇÃO DO DATASET POR ANO ===

 2020 
34571 

✓ Dataset salvo com sucesso em formato Parquet
   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/
   Estrutura: Ano=2020/, Ano=2021/, Ano=2022/, etc.


In [18]:
################################################################################
#                  VERIFICAÇÃO DO DATASET SALVO                               #
################################################################################

# Abrir dataset particionado
df_pavimentos_parquet <- open_dataset(
  "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/"
)

cat("\n=== SCHEMA DO DATASET PARQUET ===\n")
print(df_pavimentos_parquet$schema)

# Consultar amostra por ano
cat("\n=== AMOSTRA POR ANO ===\n")
anos_unicos <- df_pavimentos_parquet %>%
  select(Ano) %>%
  distinct() %>%
  collect() %>%
  pull(Ano) %>%
  sort()

for (ano in anos_unicos) {
  amostra <- df_pavimentos_parquet %>%
    filter(Ano == ano) %>%
    head(3) %>%
    collect()
  
  cat("\nAno", ano, "-", nrow(amostra), "registros (primeiros 3):\n")
  print(amostra)
}

cat("\n✓ Leitura do dataset particionado concluída com sucesso\n")

# Crie exemplo de leitura rodando um comando SELECT COUNT(*), ANO GROUP BY ANO para esse arquivo parquet.
################################################################################
#     EQUIVALENTE SQL PURO (USANDO arrow::to_duckdb)                         #
################################################################################

# Alternativa: executar SQL diretamente no dataset Parquet
library(duckdb)

# Conectar DuckDB (in-memory, zero configuração)
con <- dbConnect(duckdb::duckdb())

# Abrir dataset Parquet particionado
ds <- open_dataset("Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/")

# Registrar como tabela virtual no DuckDB
duckdb_register(con, "dados", ds)

# SELECT * LIMIT 20 (SQL puro!)
resultado <- dbGetQuery(con, "
    SELECT count(*), Ano 
    FROM dados 
    GROUP BY Ano
")

print(resultado)

# Fechar conexão
dbDisconnect(con, shutdown = TRUE)



=== SCHEMA DO DATASET PARQUET ===
Schema
Contrato: string
UF: string
Rodovia: string
km.inicial: double
km.final: double
Extensao.km: double
Data: date32[day]
Latitude: double
Longitude: double
ICC: double
ICP: double
ICM: double
Classificacao_ICM: string
Ano: int32

See $metadata for additional Schema metadata

=== AMOSTRA POR ANO ===



Ano 2020 - 3 registros (primeiros 3):
# A tibble: 3 × 14
  Contrato     UF    Rodovia km.inicial km.final Extensao.km Data       Latitude
  <chr>        <chr> <chr>        <dbl>    <dbl>       <dbl> <date>        <dbl>
1 01 00583/20… AM    BR-230        2132      212         1.2 2020-11-30  -7.20e7
2 01 00583/20… AM    BR-230         212      211         1   2020-11-30  -7.19e7
3 01 00583/20… AM    BR-230         211      210         1   2020-11-30  -7.19e7
# ℹ 6 more variables: Longitude <dbl>, ICC <dbl>, ICP <dbl>, ICM <dbl>,
#   Classificacao_ICM <chr>, Ano <int>

✓ Leitura do dataset particionado concluída com sucesso
  count_star()  Ano
1        34571 2020


## carga de Condições do Pavimento - Dezembro/ 2021


In [20]:
# Monte código R para carga do CSV conforme a seguir. Considere a formatação da data e dos números com vírgula que são floats e estão no formato brasileiro. Padronize os valore nulos para NA.
# UF;Rodovia;km inicial;km final;Extensao km;Data;Latitude;Longitude;Observação;ICC;ICP;ICM
# RS;BR-290;0,000;1,000;1,000;01/12/2021;-29,7495282;-57,0893777;Porto Seco Uruguaiana/RS.;20,000;0,000;6,000
# RS;BR-290;1,000;2,000;1,000;01/12/2021;-29,7495368;-57,089383;Porto Seco Uruguaiana/RS;20,000;0,000;6,000
# RS;BR-290;2,000;2,800;0,800;01/12/2021;-29,7495101;-57,0894646;Porto Seco Uruguaiana/RS.;20,000;0,000;6,000
# MA;BR-402;179,900;178,000;1,900;01/12/2021;-3,3311055;-43,5274917;;25,000;5,000;11,000
# MA;BR-402;178,000;177,000;1,000;01/12/2021;-3,3311055;-43,5274917;;37,500;12,500;20,000
# MA;BR-402;177,000;176,000;1,000;01/12/2021;-3,3311055;-43,5274917;;37,500;25,000;28,750
# MA;BR-402;176,000;175,000;1,000;01/12/2021;-3,3311055;-43,5274917;;37,500;5,000;14,750


################################################################################
#          CARGA DE DADOS - CONDIÇÕES DE PAVIMENTO DAS RODOVIAS              #
#                    SNV 202101A - Dezembro/2021                             #
################################################################################

# Carregar bibliotecas necessárias
library(tidyverse)
library(readr)

# Ler o CSV com separador ponto-e-vírgula
# Formato brasileiro: números com vírgula (,) como separador decimal
# Datas no formato DD/MM/YYYY
# Nota: Coluna "Observação" pode conter valores vazios (;;)
df_pavimentos_dez21 <- read.csv2(
  "Trabalho/bases-de-dados/condicoes-de-pavimento/icm-dez21.csv",
  sep = ";",
  encoding = "UTF-8",
  stringsAsFactors = FALSE,
  na.strings = c("", "-", "NA")  # Tratar vazio, hífen e NA como faltante
)

# Exibir dimensões e primeiras linhas
cat("\n=== RESUMO DO CARREGAMENTO (DEZEMBRO/2021) ===\n")
cat("Dimensões do conjunto de dados:", 
    nrow(df_pavimentos_dez21), "linhas x", 
    ncol(df_pavimentos_dez21), "colunas\n\n")

print(paste("A tabela possui", ncol(df_pavimentos_dez21), "colunas."))

colunas_com_na <- sum(colSums(is.na(df_pavimentos_dez21)) > 0)
print(paste("A tabela possui", colunas_com_na, "colunas com valores ausentes."))

cat("\n=== PRIMEIRAS 6 LINHAS ===\n")
print(head(df_pavimentos_dez21, 6))

################################################################################
#                      ESTRUTURA E TIPOS DE DADOS                             #
################################################################################

cat("\n=== ESTRUTURA DOS DADOS (ANTES DA CONVERSÃO) ===\n")
str(df_pavimentos_dez21)

################################################################################
#         CONVERSÃO E LIMPEZA DE COLUNAS NUMÉRICAS E DE DATA                  #
################################################################################

# Converter colunas numéricas (formato brasileiro com vírgula)
# Note: nomes de colunas R converte espaços em pontos
colunas_numericas <- c(
  "km.inicial", "km.final", "Extensao.km",
  "ICC", "ICP", "ICM"
)

# Aplicar conversão com tratamento de pontos decimais
for (col in colunas_numericas) {
  if (col %in% names(df_pavimentos_dez21)) {
    # Substituir vírgula por ponto e converter para numérico
    df_pavimentos_dez21[[col]] <- as.numeric(
      gsub(",", ".", gsub("\\.", "", df_pavimentos_dez21[[col]]))
    )
  }
}

# Converter coluna de Latitude e Longitude (também têm vírgula)
df_pavimentos_dez21$Latitude <- as.numeric(
  gsub(",", ".", gsub("\\.", "", df_pavimentos_dez21$Latitude))
)

df_pavimentos_dez21$Longitude <- as.numeric(
  gsub(",", ".", gsub("\\.", "", df_pavimentos_dez21$Longitude))
)

# Converter coluna Data de formato DD/MM/YYYY para Date
df_pavimentos_dez21$Data <- as.Date(df_pavimentos_dez21$Data, format = "%d/%m/%Y")

cat("\n=== ESTRUTURA DOS DADOS (APÓS CONVERSÃO) ===\n")
str(df_pavimentos_dez21)



=== RESUMO DO CARREGAMENTO (DEZEMBRO/2021) ===
Dimensões do conjunto de dados: 41414 linhas x 12 colunas

[1] "A tabela possui 12 colunas."
[1] "A tabela possui 1 colunas com valores ausentes."

=== PRIMEIRAS 6 LINHAS ===
  UF Rodovia km.inicial km.final Extensao.km       Data   Latitude Longitude
1 RS  BR-290        0.0      1.0         1.0 01/12/2021 -29.749528 -57.08938
2 RS  BR-290        1.0      2.0         1.0 01/12/2021 -29.749537 -57.08938
3 RS  BR-290        2.0      2.8         0.8 01/12/2021 -29.749510 -57.08946
4 MA  BR-402      179.9    178.0         1.9 01/12/2021  -3.331106 -43.52749
5 MA  BR-402      178.0    177.0         1.0 01/12/2021  -3.331106 -43.52749
6 MA  BR-402      177.0    176.0         1.0 01/12/2021  -3.331106 -43.52749
                 Observação  ICC  ICP   ICM
1 Porto Seco Uruguaiana/RS. 20.0  0.0  6.00
2  Porto Seco Uruguaiana/RS 20.0  0.0  6.00
3 Porto Seco Uruguaiana/RS. 20.0  0.0  6.00
4                      <NA> 25.0  5.0 11.00
5                 


=== ESTRUTURA DOS DADOS (APÓS CONVERSÃO) ===
'data.frame':	41414 obs. of  12 variables:
 $ UF         : chr  "RS" "RS" "RS" "MA" ...
 $ Rodovia    : chr  "BR-290" "BR-290" "BR-290" "BR-402" ...
 $ km.inicial : num  0 1 2 1799 178 ...
 $ km.final   : num  1 2 28 178 177 176 175 174 173 172 ...
 $ Extensao.km: num  1 1 8 19 1 1 1 1 1 1 ...
 $ Data       : Date, format: "2021-12-01" "2021-12-01" ...
 $ Latitude   : num  -2.97e+08 -2.97e+08 -2.97e+08 -3.33e+07 -3.33e+07 ...
 $ Longitude  : num  -5.71e+08 -5.71e+07 -5.71e+08 -4.35e+08 -4.35e+08 ...
 $ Observação : chr  "Porto Seco Uruguaiana/RS." "Porto Seco Uruguaiana/RS" "Porto Seco Uruguaiana/RS." NA ...
 $ ICC        : num  20 20 20 25 375 375 375 375 375 375 ...
 $ ICP        : num  0 0 0 5 125 25 5 125 5 5 ...
 $ ICM        : num  6 6 6 11 20 ...


In [22]:
# Comparando estrutura dos dois dataframes carregados (2020 vs 2021)
str(df_pavimentos)
str(df_pavimentos_dez21)

'data.frame':	34571 obs. of  14 variables:
 $ Contrato         : chr  "01 00583/2018" "01 00583/2018" "01 00583/2018" "01 00583/2018" ...
 $ UF               : chr  "AM" "AM" "AM" "AM" ...
 $ Rodovia          : chr  "BR-230" "BR-230" "BR-230" "BR-230" ...
 $ km.inicial       : num  2132 212 211 210 209 ...
 $ km.final         : num  212 211 210 209 208 214 215 198 199 200 ...
 $ Extensao.km      : num  1.2 1 1 1 1 0.8 1 0.3 1 1 ...
 $ Data             : Date, format: "2020-11-30" "2020-11-30" ...
 $ Latitude         : num  -71991363 -71949552 -71949552 -71949552 -71949552 ...
 $ Longitude        : num  -5.99e+08 -5.99e+08 -5.99e+08 -5.99e+08 -5.99e+08 ...
 $ ICC              : num  80 80 80 80 80 85 85 375 30 25 ...
 $ ICP              : num  70 70 45 70 45 70 70 325 25 25 ...
 $ ICM              : num  73 73 555 73 555 745 745 34 265 25 ...
 $ Classificacao_ICM: chr  "PÉSSIMO" "PÉSSIMO" "PÉSSIMO" "PÉSSIMO" ...
 $ Ano              : chr  "2020" "2020" "2020" "2020" ...
'data.frame':	41

In [24]:
# Carregar df_pavimentos_dez21 como uma nova partição no arquivo parquet
# "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq" particionado por ano

################################################################################
#        ADICIONAR NOVA PARTIÇÃO AO DATASET PARQUET EXISTENTE                   #
################################################################################
# Preparar dataframe: extrair ano da coluna Data
df_pavimentos_dez21 <- df_pavimentos_dez21 %>%
  mutate(
    Ano = format(Data, "%Y")
  )
# Adicionar nova partição ao dataset Parquet existente
write_dataset(
  df_pavimentos_dez21,
  path = "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq",
  format = "parquet",
  partitioning = "Ano",
  existing_data_behavior = "delete_matching"  # Delete existing partição se já existir
)
cat("\n✓ Nova partição (Dezembro/2021) adicionada com sucesso ao dataset Parquet\n")
cat("   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/\n")
cat("   Estrutura: Ano=2020/, Ano=2021/, Ano=2022/, etc.\n")
################################################################################
#                  VERIFICAÇÃO DO DATASET SALVO                               #
################################################################################
# Abrir dataset particionado
df_pavimentos_parquet <- open_dataset(
  "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/"
)   
cat("\n=== SCHEMA DO DATASET PARQUET ATUALIZADO ===\n")
print(df_pavimentos_parquet$schema)
# Consultar amostra por ano
cat("\n=== AMOSTRA POR ANO ===\n")
anos_unicos <- df_pavimentos_parquet %>%
  select(Ano) %>%
  distinct() %>%
  collect() %>%
  pull(Ano) %>%
  sort()
for (ano in anos_unicos) {
  amostra <- df_pavimentos_parquet %>%
    filter(Ano == ano) %>%
    head(3) %>%
    collect()
  cat(paste0("\nAno: ", ano, "\n"))
  print(amostra)
}
cat("\n✓ Leitura do dataset particionado atualizado concluída com sucesso\n")
# Crie exemplo de leitura rodando um comando SELECT COUNT(*), ANO GROUP BY ANO para esse arquivo parquet.
################################################################################
#     EQUIVALENTE SQL PURO (USANDO arrow::to_duckdb)
################################################################################
# Alternativa: executar SQL diretamente no dataset Parquet
library(duckdb)
# Conectar DuckDB (in-memory, zero configuração)
con <- dbConnect(duckdb::duckdb())
# Abrir dataset Parquet particionado
ds <- open_dataset("Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/")
# Registrar como tabela virtual no DuckDB
duckdb_register(con, "dados", ds)
# SELECT COUNT(*), Ano GROUP BY Ano (SQL puro!)
resultado <- dbGetQuery(con, "
    SELECT count(*) AS Total_Registros, Ano 
    FROM dados 
    GROUP BY Ano
") 
print(resultado)
# Fechar conexão
dbDisconnect(con, shutdown = TRUE)



✓ Nova partição (Dezembro/2021) adicionada com sucesso ao dataset Parquet
   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/
   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/
   Estrutura: Ano=2020/, Ano=2021/, Ano=2022/, etc.
   Estrutura: Ano=2020/, Ano=2021/, Ano=2022/, etc.

=== SCHEMA DO DATASET PARQUET ATUALIZADO ===

=== SCHEMA DO DATASET PARQUET ATUALIZADO ===
Schema
Contrato: string
UF: string
Rodovia: string
km.inicial: double
km.final: double
Extensao.km: double
Data: date32[day]
Latitude: double
Longitude: double
ICC: double
ICP: double
ICM: double
Classificacao_ICM: string
Ano: int32

See $metadata for additional Schema metadata

=== AMOSTRA POR ANO ===
Schema
Contrato: string
UF: string
Rodovia: string
km.inicial: double
km.final: double
Extensao.km: double
Data: date32[day]
Latitude: double
Longitude: double
ICC: double
ICP: double
ICM: double
Classificacao_ICM: string
Ano: int32

See $metadata for a


Ano: 2020
# A tibble: 3 × 14
  Contrato     UF    Rodovia km.inicial km.final Extensao.km Data       Latitude
  <chr>        <chr> <chr>        <dbl>    <dbl>       <dbl> <date>        <dbl>
1 01 00583/20… AM    BR-230        2132      212         1.2 2020-11-30  -7.20e7
2 01 00583/20… AM    BR-230         212      211         1   2020-11-30  -7.19e7
3 01 00583/20… AM    BR-230         211      210         1   2020-11-30  -7.19e7
# ℹ 6 more variables: Longitude <dbl>, ICC <dbl>, ICP <dbl>, ICM <dbl>,
#   Classificacao_ICM <chr>, Ano <int>

Ano: 2021
# A tibble: 3 × 14
  Contrato UF    Rodovia km.inicial km.final Extensao.km Data         Latitude
  <chr>    <chr> <chr>        <dbl>    <dbl>       <dbl> <date>          <dbl>
1 NA       RS    BR-290           0        1           1 2021-12-01 -297495282
2 NA       RS    BR-290           1        2           1 2021-12-01 -297495368
3 NA       RS    BR-290           2       28           8 2021-12-01 -297495101
# ℹ 6 more variables: Longitu

Conseguimos carregar os anos de 2020 e 2021 sem problemas, pois a estrutura dos dataframes é a mesma.

In [3]:
# Vamos padronizar o arquivo no formato Rdata, vamos ler o parquet e salvar em Rdata
################################################################################
#        CONVERSÃO DO DATASET PARQUET PARA RDATA                             #
################################################################################
# Carregar biblioteca arrow (se não estiver carregada)
library(arrow)
# Ler dataset Parquet particionado
df_pavimentos_parquet <- open_dataset(
  "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento-pq/"
) %>%
  collect()  # Coletar todos os dados em um dataframe R
# Salvar dataframe como arquivo Rdata
save(df_pavimentos_parquet, 
     file = "Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento.Rdata")
cat("\n✓ Dataset salvo com sucesso em formato Rdata\n")
cat("   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento.Rdata\n")


✓ Dataset salvo com sucesso em formato Rdata
   Caminho: Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento.Rdata


In [4]:
# Visualizar o conteúdo das primeiras 20 linhas do arquivo Rdata
load("Trabalho/bases-de-dados/condicoes-de-pavimento/condicoes-de-pavimento.Rdata")
head(df_pavimentos_parquet, 20)

Contrato      UF Rodovia km.inicial km.final Extensao.km Data      
1  01 00583/2018 AM BR-230  2132       212      1.2         2020-11-30
2  01 00583/2018 AM BR-230   212       211      1.0         2020-11-30
3  01 00583/2018 AM BR-230   211       210      1.0         2020-11-30
4  01 00583/2018 AM BR-230   210       209      1.0         2020-11-30
5  01 00583/2018 AM BR-230   209       208      1.0         2020-11-30
6  01 00583/2018 AM BR-230  2132       214      0.8         2020-11-30
7  01 00583/2018 AM BR-230   214       215      1.0         2020-11-30
8  09 00346/2018 PR BR-476  1977       198      0.3         2020-11-05
9  09 00346/2018 PR BR-476   198       199      1.0         2020-11-05
10 09 00346/2018 PR BR-476   199       200      1.0         2020-11-05
11 09 00346/2018 PR BR-476   200       201      1.0         2020-11-05
12 09 00346/2018 PR BR-476   201       202      1.0         2020-11-05
13 09 00346/2018 PR BR-476   202       203      1.0         2020-11-05
14 09 00346/2018 PR BR-476   203       204      1.0         2020-11-05
15 09 00346/2018 PR BR-476   204       205      1.0         2020-11-05
16 09 00346/2018 PR BR-476   205       206      1.0         2020-11-05
17 09 00346/2018 PR BR-476   206       207      1.0         2020-11-05
18 09 00346/2018 PR BR-476   207       208      1.0         2020-11-05
19 09 00346/2018 PR BR-476   208       209      1.0         2020-11-05
20 09 00346/2018 PR BR-476   209       210      1.0         2020-11-05
   Latitude   Longitude  ICC ICP ICM  Classificacao_ICM Ano 
1   -71991363 -598909876  80  70   73 PÉSSIMO           2020
2   -71949552 -598855125  80  70   73 PÉSSIMO           2020
3   -71949552 -598855125  80  45  555 PÉSSIMO           2020
4   -71949552 -598855125  80  70   73 PÉSSIMO           2020
5   -71949552 -598855125  80  45  555 PÉSSIMO           2020
6   -71788378  -59846564  85  70  745 PÉSSIMO           2020
7   -71923246 -598833152  85  70  745 PÉSSIMO           2020
8  -257656242 -497390803 375 325   34 REGULAR           2020
9  -257604123 -497378996  30  25  265 PÉSSIMO           2020
10 -257729066 -497459091  25  25   25 BOM               2020
11 -257729066 -497459091 625 475   52 RUIM              2020
12 -257754729 -497657167  85 475 5875 PÉSSIMO           2020
13 -257764558 -497752899  45 475 4675 PÉSSIMO           2020
14 -257786094 -497855551  50 475 4825 PÉSSIMO           2020
15 -257786094 -497855551  50 325 3775 PÉSSIMO           2020
16 -257794613 -498046225 425 475   46 REGULAR           2020
17 -257794613 -498046225 425 475   46 REGULAR           2020
18 -257794613 -498046225 375 475  445 PÉSSIMO           2020
19 -257794613 -498046225 625 475   52 RUIM              2020
20 -257794613 -498046225  45 475 4675 PÉSSIMO           2020

# Vamos criar um dicionário de dados atualizado para a amostra abaixo, usar o format markdown

Contrato	UF	Rodovia	km.inicial	km.final	Extensao.km	Data	Latitude	Longitude	ICC	ICP	ICM	Classificacao_ICM	Ano
<chr>	<chr>	<chr>	<dbl>	<dbl>	<dbl>	<date>	<dbl>	<dbl>	<dbl>	<dbl>	<dbl>	<chr>	<int>
01 00583/2018	AM	BR-230	2132	212	1.2	2020-11-30	-71991363	-598909876	80	70	73	PÉSSIMO	2020
01 00583/2018	AM	BR-230	212	211	1.0	2020-11-30	-71949552	-598855125	80	70	73	PÉSSIMO	2020
01 00583/2018	AM	BR-230	211	210	1.0	2020-11-30	-71949552	-598855125	80	45	555	PÉSSIMO	2020
01 00583/2018	AM	BR-230	210	209	1.0	2020-11-30	-71949552	-598855125	80	70	73	PÉSSIMO	2020
01 00583/2018	AM	BR-230	209	208	1.0	2020-11-30	-71949552	-598855125	80	45	555	PÉSSIMO	2020
01 00583/2018	AM	BR-230	2132	214	0.8	2020-11-30	-71788378	-59846564	85	70	745	PÉSSIMO	2020
01 00583/2018	AM	BR-230	214	215	1.0	2020-11-30	-71923246	-598833152	85	70	745	PÉSSIMO	2020
09 00346/2018	PR	BR-476	1977	198	0.3	2020-11-05	-257656242	-497390803	375	325	34	REGULAR	2020
09 00346/2018	PR	BR-476	198	199	1.0	2020-11-05	-257604123	-497378996	30	25	265	PÉSSIMO	2020
09 00346/2018	PR	BR-476	199	200	1.0	2020-11-05	-257729066	-497459091	25	25	25	BOM	2020
09 00346/2018	PR	BR-476	200	201	1.0	2020-11-05	-257729066	-497459091	625	475	52	RUIM	2020
09 00346/2018	PR	BR-476	201	202	1.0	2020-11-05	-257754729	-497657167	85	475	5875	PÉSSIMO	2020
09 00346/2018	PR	BR-476	202	203	1.0	2020-11-05	-257764558	-497752899	45	475	4675	PÉSSIMO	2020
09 00346/2018	PR	BR-476	203	204	1.0	2020-11-05	-257786094	-497855551	50	475	4825	PÉSSIMO	2020
09 00346/2018	PR	BR-476	204	205	1.0	2020-11-05	-257786094	-497855551	50	325	3775	PÉSSIMO	2020
09 00346/2018	PR	BR-476	205	206	1.0	2020-11-05	-257794613	-498046225	425	475	46	REGULAR	2020
09 00346/2018	PR	BR-476	206	207	1.0	2020-11-05	-257794613	-498046225	425	475	46	REGULAR	2020
09 00346/2018	PR	BR-476	207	208	1.0	2020-11-05	-257794613	-498046225	375	475	445	PÉSSIMO	2020
09 00346/2018	PR	BR-476	208	209	1.0	2020-11-05	-257794613	-498046225	625	475	52	RUIM	2020
09 00346/2018	PR	BR-476	209	210	1.0	2020-11-05	-257794613	-498046225	45	475	4675	PÉSSIMO	2020

# Dicionário de Dados – Condições de Pavimento das Rodovias Brasileiras

## Visão Geral
Dataset contendo avaliações de índices de condição de pavimento em rodovias federais brasileiras, com registros segmentados por quilometragem e data de inspeção.

---

## Definição dos Campos

| Campo | Tipo | Formato | Descrição | Exemplo | Observações |
|---|---|---|---|---|---|
| **Contrato** | Caractere | `NN NNNNN/AAAA` | Número do contrato de serviço de manutenção/inspeção | `01 00583/2018` | Identifica o contrato responsável pela rodovia |
| **UF** | Caractere | `XX` | Sigla da Unidade Federativa (estado) | `AM`, `PR` | Formato ISO 3166-2 BR (ex: AM=Amazonas, PR=Paraná) |
| **Rodovia** | Caractere | `XX-NNN` | Designação da rodovia federal | `BR-230`, `BR-476` | Formato padrão DNIT |
| **km.inicial** | Numérico (decimal) | `NNNN,N` | Quilometragem inicial do segmento avaliado | `212,0` | Localiza o ponto de início da inspeção |
| **km.final** | Numérico (decimal) | `NNNN,N` | Quilometragem final do segmento avaliado | `211,0` | Localiza o ponto final da inspeção |
| **Extensao.km** | Numérico (decimal) | `N,NNN` | Extensão do segmento em quilômetros (diferença entre km.final e km.inicial) | `1,0` | Valor absoluto; segmentos típicos de 0,3–1,2 km |
| **Data** | Data | `AAAA-MM-DD` | Data da inspeção mais recente | `2020-11-30` | Formato ISO 8601 após conversão (original: DD/MM/YYYY) |
| **Latitude** | Numérico (decimal) | `±NN,NNNNNNN` | Coordenada geográfica de latitude (datum WGS84) | `-7,1991363` | Escala em graus decimais; negativo = Sul |
| **Longitude** | Numérico (decimal) | `±NN,NNNNNNN` | Coordenada geográfica de longitude (datum WGS84) | `-59,8909876` | Escala em graus decimais; negativo = Oeste |
| **ICC** | Numérico (decimal) | `NN,NNN` | **Índice de Condição da Conservação** (manutenção estrutural) | `80,0`, `25,0` | Escala 0–100; maior = melhor condição |
| **ICP** | Numérico (decimal) | `NN,NNN` | **Índice de Condição do Pavimento** (desgaste superficial) | `70,0`, `25,0` | Escala 0–100; maior = melhor condição |
| **ICM** | Numérico (decimal) | `NN,NNN` | **Índice de Condição de Manutenção** (composição ponderada) | `73,0`, `25,0` | Combinação de ICC e ICP; determina classificação |
| **Classificacao_ICM** | Caractere | `TEXTO` | Categoria qualitativa baseada no valor de ICM | `BOM`, `REGULAR`, `RUIM`, `PÉSSIMO` | Ver matriz de classificação abaixo |
| **Ano** | Inteiro | `AAAA` | Ano de extração da data de inspeção | `2020` | Derivado de **Data**; facilita particionamento |

---

## Matriz de Classificação de ICM

| Classificação | Intervalo ICM | Interpretação |
|---|---|---|
| **BOM** | ICM < 30 | Pavimento em excelente estado; manutenção preventiva suficiente |
| **REGULAR** | 30 ≤ ICM < 50 | Pavimento aceitável; requer manutenção rotineira em breve |
| **RUIM** | 50 ≤ ICM < 70 | Pavimento degradado; manutenção urgente recomendada |
| **PÉSSIMO** | ICM ≥ 70 | Pavimento crítico; restauração estrutural necessária |

---

## Notas Técnicas

1. **Fonte de Dados**: DNIT (Departamento Nacional de Infraestrutura de Transportes) – Sistema Nacional de Viações (SNV).

2. **Valores Faltantes**: Codificados como `NA` após limpeza; colunas sem dados válidos são raras (geralmente observações descritivas opcionais).

3. **Coordenadas Geográficas**: 
   - Sistema de referência: WGS84 (EPSG:4326)
   - Formato armazenado: Graus decimais (não DMS)
   - Aplicações: Georeferenciamento em GIS, cálculo de distâncias

4. **Particionamento**: Dataset organizado em estrutura Parquet com partições por **Ano**, otimizando consultas ano-a-ano.

5. **Precisão Numérica**: Índices (ICC, ICP, ICM) frequentemente apresentam 1–3 casas decimais em brutos; recomenda-se arredondamento para 2 casas em relatórios.

---

## Estrutura de Armazenamento

- **Formato**: Parquet (particionado por Ano) + RData
- **Localização**: condicoes-de-pavimento
- **Partições**: `Ano=2020/`, `Ano=2021/`, etc.